[OpenMP] Port the OpenMP device runtime to direct C++ compilation

jhuber6 opened the pull request:
Summary:
This removes the use of OpenMP offloading to build the device runtime.
The main benefit here is that we no longer need to rely on offloading
semantics to build a device only runtime. Things like variants are now
no longer needed and can just be simple if-defs. In the future, I will
remove most of the special handling here and fold it into calls to the
`<gpuintrin.h>` functions instead. Additionally I will rework the
compilation to make this a separate runtime.

The current plan is to have this, but make including OpenMP and
offloading either automatically add it, or print a warning if it's
missing. This will allow us to use a normal CMake workflow and delete
all the weird 'lets pull the clang binary out of the build' business.
```
-DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_RUNTIMES=offload
-DLLVM_RUNTIME_TARGETS=amdgcn-amd-amdhsa
```

After that, linking the OpenMP device runtime will be `-Xoffload-linker
-lomp`. I.e. no more fat binary business.

Only look at the most recent commit since this includes the two dependencies
(fix to AMDGPUEmitPrintfBinding and the PointerToMember bug).

jhuber6 committed:
[OpenMP] Port the OpenMP device runtime to direct C++ compilation

Summary:
This removes the use of OpenMP offloading to build the device runtime.
The main benefit here is that we no longer need to rely on offloading
semantics to build a device only runtime. Things like variants are now
no longer needed and can just be simple if-defs. In the future, I will
remove most of the special handling here and fold it into calls to the
`<gpuintrin.h>` functions instead. Additionally I will rework the
compilation to make this a separate runtime.

The current plan is to have this, but make including OpenMP and
offloading either automatically add it, or print a warning if it's
missing. This will allow us to use a normal CMake workflow and delete
all the weird 'lets pull the clang binary out of the build' business.
```
-DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_RUNTIMES=offload
-DLLVM_RUNTIME_TARGETS=amdgcn-amd-amdhsa
```

After that, linking the OpenMP device runtime will be `-Xoffload-linker
-lomp`. I.e. no more fat binary business.

shiltian commented:
I like it but we need to test it on both AMD and NVIDIA cards. This'd be a topic for next meeting.

jhuber6 commented:
Works AFIACT once I stripped out the buggy member to ptr handling, but definitely needs more tests and hopefull the other two dependencies can be landed uncontroversially.

Meinersbur commented on offload/DeviceRTL/include/State.h:
```diff
@@ -88,8 +86,7 @@ struct TeamStateTy {
 
-extern TeamStateTy TeamState;
-#pragma omp allocate(TeamState) allocator(omp_pteam_mem_alloc)
+extern TeamStateTy [[clang::address_space(3)]] TeamState;
```
Add comments what address space `3` is?

jhuber6 replied:
Eventually I'm going to port this to use `<gpuintrin.h>` while defined `__gpu_local` which is more descriptive. For now I can add a comment.

Meinersbur commented on offload/DeviceRTL/src/Debug.cpp:
```diff
@@ -21,8 +21,6 @@
 
 using namespace ompx;
```
If `printf` is meant to resolve to `ompx::printf`, I would prefer removing `using namespace ompx;`, so it becomes  clear this is not the standard `printf`. (or give it a different name)

jhuber6 replied:
I could put it in another namespace.

Meinersbur commented on offload/DeviceRTL/src/Tasking.cpp:
```diff
@@ -20,16 +20,14 @@
                                         TaskFnTy TaskFn) {
   auto TaskSizeInclPrivateValuesPadded =
-      utils::roundUp(TaskSizeInclPrivateValues, uint64_t(sizeof(void *)));
+      utils::roundUp(TaskSizeInclPrivateValues, size_t(sizeof(void *)));
```
```suggestion
      utils::roundUp(TaskSizeInclPrivateValues, sizeof(void *));
```
`sizeof`, [by defintion](https://en.cppreference.com/w/cpp/language/sizeof), returns `size_t`

Meinersbur commented on offload/DeviceRTL/src/DeviceUtils.cpp:
```diff
@@ -15,14 +15,10 @@
 
 namespace impl {
 
-bool isSharedMemPtr(const void *Ptr) { return false; }
```
Doesn't any no-device fallback code still need an implemenations of such functions?

jhuber6 replied:
This is pure device code, this only exists to make the variant system happy because it still needs to parse for the CPU compilation. This completely removes that inconvenient requirement.

Meinersbur reviewed (commented):
Is the plan to eventually make CMake handle `--target=${target_triple}` (and include dependencies)? like it does with `LLVM_ENABLE_RUNTIMES`?

I assume header files that are #included in user programs (e.g. `__clang_openmp_device_functions.h`) wil remain unaffected?

jhuber6 commented:
> Is the plan to eventually make CMake handle `--target=${target_triple}` (and include dependencies)? like it does with `LLVM_ENABLE_RUNTIMES`?
> 
> I assume header files that are #included in user programs (e.g. `__clang_openmp_device_functions.h`) wil remain unaffected?

Yes, the idea is to remove all of the hacky compilations and just use `add_library` or `add_executable`. That'll make the build more 'normal'.

jplehr commented:
I like where this is going to remove all the CMake trickery we currently have and simply have this as a runtime build.

jhuber6 commented:
@sarnex FYI, my current plan is to do this and then port all of the target dependent stuff to hopefully be in `<gpuintrin.h>`. Ideally we can then create a `sprivintrin.h` header that defines similar stuff and pick up most of this for free.

sarnex commented:
Cool, thanks! We have some inklings of the required changes for the SPIR-V Device RTL internally, so once we have a target-specific header we can modify we can start contributing those.

jplehr commented:
I tested this patch with the CMake cache file in `offload/cmake/caches/AMDGPUBot.cmake` and see errors about some nvptx that is built. The cache only enables `host;AMDGPU`. Does this patch miss / remove guard?

jhuber6 commented:
> I tested this patch with the CMake cache file in `offload/cmake/caches/AMDGPUBot.cmake` and see errors about some nvptx that is built. The cache only enables `host;AMDGPU`. Does this patch miss / remove guard?

I didn't see issues locally, if you run `ninja check-offload` there's always like 12 NVPTX tests that fail for me with and without this patch. Could you copy them here?

jhuber6 commented:
@ronlieb Downstream application https://gist.github.com/jhuber6/ebf360f76158ea4cffc38dc925ff160d
@jplehr PTAL since it passes upstream testing.

jplehr commented:
I assume that this requires changes to the buildbot configs?

jplehr commented:
Do you plan to have a follow-up that addresses the concerns about the explicit address space numbers? I too think it would be easier to read if some sort of enum is used instead of plain numbers.

jhuber6 commented:
> Do you plan to have a follow-up that addresses the concerns about the explicit address space numbers? I too think it would be easier to read if some sort of enum is used instead of plain numbers.

Yes, I'm going to port this stuff to use `gpuintrin.h` instead so future architectures just need to define that header's definitions.

jplehr reviewed (approved):
LGTM
